velox/exec/MergingVectorInput.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MergingVectorInput.h"
+#include "velox/exec/OperatorUtils.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorInput::MergingVectorInput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int32_t maxMergingRows,
+    bool preserveOrder)
+    : pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      maxMergingRows_(maxMergingRows),
+      preserveOrder_(preserveOrder) {}
+
+RowVectorPtr MergingVectorInput::getVector(bool noMoreInput) {
+  if (noMoreInput) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    const auto res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorInput::addVector(RowVectorPtr vector) {
+  if (nullptr == vector) {
+    return;
+  }
+
+  const auto numInput = vector->size();
+  if (numInput == 0) {
+    return;
+  }
+
+  // Avoid memory copying for pages that are big enough
+  if (numInput >= maxMergingRows_) {
+    if (preserveOrder_) {
+      // In order to preserve the order of the rows, we need to flush first.
+      flush();
+    }
+    outputQueue_.push(std::move(vector));
+    return;
+  }
+
+  buffer(std::move(vector));
+}
+
+void MergingVectorInput::buffer(RowVectorPtr vector) {
+  const auto numInput = vector->size();
+
+  // In order to prevent vector.resize() cause memory copy, the initial value
+  // of bufferInputs_ row count is preferredOutputBatchRows_ and cannot exceed
+  // preferredOutputBatchRows_.
+  if (numBufferRows_ + numInput > preferredOutputBatchRows_) {
+    flush();
+  }
+
+  if (!bufferInputs_) {
```

Comment:
Perhaps, we could defer copying until we get a second vector that we want to combine with already buffered vector. Otherwise, we may end up copying unnecessarily if second vector will keep hitting the check on L75.

Reply:
```
can we simply implement as
if (bufferInputs_ == nullptr) {
  bufferedInput_ = std::move(vector);
}
```

Reply:
If the input is a lazy vector, it seems we can't defer copying until the second vector.

Reply:
True. We can load the lazy, but we do not need to copy it.

Reply:
But if we load the lazy vector here,  it's ok when downstream operator is aggregation, but it not good for that the downstream operator no need to load the lazy vector.

Reply:
@shenh062326 Currently, you are copying the vector. I suggest to instead load lazy. Either way lazy vector cannot survive MergingVectorInput, but loading lazy is cheaper than loading and copying.

Reply:
Done.